fuzz/ml-kem.c

```diff
@@ -0,0 +1,685 @@
+ * @param unused Unused parameter (reserved for future use).
+ *
+ * @note The generated key is allocated using OpenSSL's EVP_PKEY functions
+ *       and should be freed using `EVP_PKEY_free()`.
+ */
+static void keygen_mlkem_real_key(uint8_t **buf, size_t *len,
+                                  void **key1, void **key2)
+{
+    char *keytype = NULL;
+    size_t keylen = 0;
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY **key;
+
+    *key1 = *key2 = NULL;
+
+    key = (EVP_PKEY **)key1;
+
+again:
+    /*
+     * Only generate valid key types and lengths
+     */
+    if (!select_keytype_and_size(buf, len, &keytype, &keylen, 1))
+        return;
+
+    ctx = EVP_PKEY_CTX_new_from_name(NULL, keytype, NULL);
+    if (!ctx) {
+        fprintf(stderr, "Failed to generate ctx\n");
+        return;
+    }
+
+    if (!EVP_PKEY_keygen_init(ctx)) {
+        fprintf(stderr, "Failed to init keygen ctx\n");
+        goto err;
+    }
+
+    *key = EVP_PKEY_new();
+    if (*key == NULL)
+        goto err;
+
+    if (!EVP_PKEY_generate(ctx, key)) {
+        fprintf(stderr, "Failed to generate new real key\n");
+        goto err;
+    }
+
+    if (key == (EVP_PKEY **)key1) {
+        EVP_PKEY_CTX_free(ctx);
+        key = (EVP_PKEY **)key2;
+        goto again;
+    }
+
+err:
+    EVP_PKEY_CTX_free(ctx);
+    return;
+}
+
+/**
+ * @brief Performs key encapsulation and decapsulation using an EVP_PKEY.
+ *
+ * This function generates a random key, encapsulates it using the provided
+ * public key, then decapsulates it to retrieve the original key. It makes
+ * use of OpenSSL's EVP_PKEY API for encryption and decryption.
+ *
+ * @param[out] buf   Unused output buffer (reserved for future use).
+ * @param[out] len   Unused length parameter (reserved for future use).
+ * @param[in]  key1  Pointer to an EVP_PKEY structure used for key operations.
+ * @param[in]  in2   Unused input parameter (reserved for future use).
+ * @param[out] out1  Unused output parameter (reserved for future use).
+ * @param[out] out2  Unused output parameter (reserved for future use).
+ */
+static void mlkem_encap_decap(uint8_t **buf, size_t *len, void *key1, void *in2,
+                              void **out1, void **out2)
+{
+    EVP_PKEY *key = (EVP_PKEY *)key1;
+    EVP_PKEY_CTX *ctx;
+    unsigned char genkey[128];
+    size_t genkey_len = 128;
+    unsigned char unwrappedkey[2048];
+    size_t unwrappedkey_len = 2048;
+    unsigned char wrapkey[2048];
+    size_t wrapkey_len = 2048;
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, NULL);
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed to allocate ctx\n");
+        goto err;
+    }
+
+    if (!EVP_PKEY_encapsulate_init(ctx, NULL)) {
+        fprintf(stderr, "Failed to init encap context\n");
+        goto err;
+    }
+
+    if (!RAND_bytes(genkey, genkey_len))
+        goto err;
+
+    if (EVP_PKEY_encapsulate(ctx, wrapkey, &wrapkey_len, genkey, &genkey_len) <= 0) {
```

Comment:
FYI (not a change suggestion): The shared secret size for `ML-KEM` (`genkey_len`) is always 32 bytes, and `wrapkey_len` tops out at 1568, FWIW.